File: server/src/repos/password_reset.rs
use chrono::{DateTime, Utc};
use sqlx::PgPool;
use uuid::Uuid;

use crate::db::schema;

#[derive(Debug, sqlx::FromRow)]
pub struct PasswordResetTokenRow {
    pub id: Uuid,
    pub user_id: Uuid,
    pub token_hash: Vec<u8>,
    pub expires_at: DateTime<Utc>,
    pub used_at: Option<DateTime<Utc>>,
}

/// Removes outstanding reset tokens for this user, then inserts a new one.
pub async fn replace_token_for_user(
    pool: &PgPool,
    user_id: Uuid,
    token_hash: &[u8],
    expires_at: DateTime<Utc>,
) -> Result<(), sqlx::Error> {
    let mut tx = pool.begin().await?;
    sqlx::query(&format!(
        r#"DELETE FROM {} WHERE user_id = $1"#,
        schema::PASSWORD_RESET_TOKENS
    ))
    .bind(user_id)
    .execute(&mut *tx)
    .await?;

    sqlx::query(&format!(
        r#"
        INSERT INTO {} (user_id, token_hash, expires_at)
        VALUES ($1, $2, $3)
        "#,
        schema::PASSWORD_RESET_TOKENS
    ))
    .bind(user_id)
    .bind(token_hash)
    .bind(expires_at)
    .execute(&mut *tx)
    .await?;

    tx.commit().await?;
    Ok(())
}

pub async fn find_by_token_hash(
    pool: &PgPool,
    token_hash: &[u8],
) -> Result<Option<PasswordResetTokenRow>, sqlx::Error> {
    sqlx::query_as::<_, PasswordResetTokenRow>(&format!(
        r#"
        SELECT id, user_id, token_hash, expires_at, used_at
        FROM {}
        WHERE token_hash = $1
        "#,
        schema::PASSWORD_RESET_TOKENS
    ))
    .bind(token_hash)
    .fetch_optional(pool)
    .await
}

pub async fn mark_used_and_set_password(
    pool: &PgPool,
    token_id: Uuid,
    user_id: Uuid,
    password_hash: &str,
) -> Result<bool, sqlx::Error> {
    let mut tx = pool.begin().await?;
    let n = sqlx::query(&format!(
        r#"
        UPDATE {}
        SET used_at = NOW()
        WHERE id = $1 AND user_id = $2 AND used_at IS NULL
        "#,
        schema::PASSWORD_RESET_TOKENS
    ))
    .bind(token_id)
    .bind(user_id)
    .execute(&mut *tx)
    .await?
    .rows_affected();
    if n != 1 {
        tx.rollback().await?;
        return Ok(false);
    }

    sqlx::query(&format!(
        r#"UPDATE {} SET password_hash = $2 WHERE id = $1"#,
        schema::USERS
    ))
    .bind(user_id)
    .bind(password_hash)
    .execute(&mut *tx)
    .await?;

    tx.commit().await?;
    Ok(true)
}

File: server-new/internal/repos/passwordreset/passwordreset.go
// Package passwordreset ports server/src/repos/password_reset.rs.
package passwordreset

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// TokenRow is a row from password_reset_tokens.
type TokenRow struct {
	ID        string
	UserID    string
	TokenHash []byte
	ExpiresAt time.Time
	UsedAt    *time.Time
}

// ReplaceTokenForUser deletes any existing reset tokens for the user and inserts a new one.
func ReplaceTokenForUser(ctx context.Context, pool *pgxpool.Pool, userID uuid.UUID, tokenHash []byte, expiresAt time.Time) error {
	tx, err := pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	_, err = tx.Exec(ctx, `DELETE FROM "user".password_reset_tokens WHERE user_id = $1`, userID.String())
	if err != nil {
		return err
	}
	_, err = tx.Exec(ctx, `INSERT INTO "user".password_reset_tokens (user_id, token_hash, expires_at) VALUES ($1::uuid, $2, $3)`,
		userID.String(), tokenHash, expiresAt,
	)
	if err != nil {
		return err
	}
	return tx.Commit(ctx)
}

// FindByTokenHash looks up a reset token by SHA-256 hash of the raw token.
func FindByTokenHash(ctx context.Context, pool *pgxpool.Pool, tokenHash []byte) (*TokenRow, error) {
	const q = `SELECT id::text, user_id::text, token_hash, expires_at, used_at FROM "user".password_reset_tokens WHERE token_hash = $1`
	var r TokenRow
	err := pool.QueryRow(ctx, q, tokenHash).Scan(&r.ID, &r.UserID, &r.TokenHash, &r.ExpiresAt, &r.UsedAt)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &r, nil
}

// MarkUsedAndSetPassword marks the token used and sets the user password. Returns false if the token was invalid or already used.
func MarkUsedAndSetPassword(ctx context.Context, pool *pgxpool.Pool, tokenID, userID uuid.UUID, passwordHash string) (bool, error) {
	tx, err := pool.Begin(ctx)
	if err != nil {
		return false, err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	tag, err := tx.Exec(ctx, `
UPDATE "user".password_reset_tokens
SET used_at = NOW()
WHERE id = $1::uuid AND user_id = $2::uuid AND used_at IS NULL
`, tokenID.String(), userID.String())
	if err != nil {
		return false, err
	}
	if tag.RowsAffected() != 1 {
		return false, nil
	}

	_, err = tx.Exec(ctx, `UPDATE "user".users SET password_hash = $2 WHERE id = $1::uuid`, userID.String(), passwordHash)
	if err != nil {
		return false, err
	}
	if err := tx.Commit(ctx); err != nil {
		return false, err
	}
	return true, nil
}
